### PR TITLE
Feature/ticket146

### PR DESCRIPTION
**Pull request by rcbjBlueMars**
## Description
Please include a summary of the change, motivation and context.

* Adding complete set of default values.
* Simplified logic in AgentConfig class.
* Check for unknown attributes. Print error, but continue processing.
* Unittests for loading config file and environment variables.

### Testing
*Standard CICD pipeline checks.
* New set of AgentConfig unittests.
* 
### Checklist:
- [ X] My changes generate no new warnings
- [ X] I have added tests that prove my fix is effective or that my feature works
- [ X] Any dependent changes have been merged and published in downstream modules

### Documentation
* pydocs has been regenerated.

<!--

### CODE REVIEW

File: agent-config.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+service_name: pythonagent_001
```
Comment (jcchavezs) [inline agent-config.yaml:1]: I am curious about this file, do we need this? I see there is another one for testing so not sure we should include this. Could you elaborate a bit @rcbjBlueMars ?

File: agent-config.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  http_body:
+    request: false
+    response: false
+  rpc_metadata:
+    request: false
+    response: false
+  rpc_body:
+    request: false
+    response: false
+propagation_formats: "B3"
```
Comment (jcchavezs) [inline agent-config.yaml:18]: I think this should be a list.

File: src/hypertrace/agent/__init__.py [hunk 3]
```diff
@@ -172,11 +172,11 @@ def register_processor(self, processor): # pylint: disable=R1710
         if self.is_enabled():
             return self._init.register_processor(processor)
 
-    def get_config(self):
-        '''Return configuration object'''
-        return self._config
+#    def get_config(self) -> r:
```
Comment (jcchavezs) [inline src/hypertrace/agent/__init__.py:175]: Shall we remove this?

File: src/hypertrace/agent/__init__.py [hunk 4]
```diff
@@ -172,11 +172,11 @@ def register_processor(self, processor): # pylint: disable=R1710
...
+#        '''Return configuration object'''
+#        return self._config.get_config()
 
-    def is_enabled(self): # pylint: disable=R0201
+    def is_enabled(self) -> bool: # pylint: disable=R0201
```
Comment (jcchavezs) [inline src/hypertrace/agent/__init__.py:179]: +1 I think we should include return type as much as possible

File: src/hypertrace/agent/config/__init__.py [hunk 5]
```diff
@@ -110,169 +103,227 @@ def __init__(self):  # pylint: disable=R0912,R0915
 
         if 'HT_REPORTING_TOKEN' in os.environ:
             logger.debug("[env] Loaded HT_REPORTING_TOKEN from env")
-            reporting_token = os.environ['HT_REPORTING_TOKEN']
+            self.config['reporting']['token'] = os.environ['HT_REPORTING_TOKEN']
```
Comment (jcchavezs) [inline src/hypertrace/agent/config/__init__.py:106]: Shall we check `self.config['reporting']` exist?

File: src/hypertrace/agent/config/__init__.py [hunk 6]
```diff
@@ -110,169 +103,227 @@ def __init__(self):  # pylint: disable=R0912,R0915
...
+                    self.validate_config_elements(config_element[key], \
+                      eval('agent_config_base.' + key)) # pylint: disable=W0123
+                    continue
+                except AttributeError as err:
+                    logger.error('Unknown attribute encountered: exception=%s, stacktrace=%s',
+                                 err,
+                                 traceback.format_exc())
+                    continue
+            if isinstance(config_element[key], (str, bool, int, list)):
+                logger.debug('is string')
```
Comment (jcchavezs) [inline src/hypertrace/agent/config/__init__.py:300]: sounds contradictory

File: src/hypertrace/agent/config/agent-config.yaml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+service_name: pythonagent_001
```
Comment (jcchavezs) [inline src/hypertrace/agent/config/agent-config.yaml:1]: Is this file needed? Could you please elaborate on why we need to check in this?

File: src/hypertrace/agent/config/test_unittests.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
+'''Unittest for merging config entries.'''
```
Comment (jcchavezs) [inline src/hypertrace/agent/config/test_unittests.py:1]: I think a good convention would be `test_<entity_we_are_testing>`.py

File: src/hypertrace/agent/config/test_unittests.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+import os
+from hypertrace.agent.config import AgentConfig
+from . import merge_config
+from . import load_config_from_file
+from . import DEFAULT_AGENT_CONFIG
+
+
+def test_merge_config():
+    '''Unittest functionx for merging config entries.'''
```
Comment (jcchavezs) [inline src/hypertrace/agent/config/test_unittests.py:10]: I am not sure we need this doc for tests but function name should describe what we are testing.

File: src/hypertrace/agent/config/test_unittests.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+    assert config.agent_config.data_capture.body_max_size_bytes == 131072
+    assert config.agent_config.propagation_formats == 0
+    assert not config.agent_config.enabled
+    assert config.agent_config.resource_attributes == {'service_name': 'pythonagent_001'}
+
+
+def test_env_config():
+    '''Unittest functionx for env config entries.'''
+    print('Initializing agent.')
+    os.environ["HT_SERVICE_NAME"] = "pythonagent_002"
```
Comment (jcchavezs) [inline src/hypertrace/agent/config/test_unittests.py:73]: We should deinit the env vars after the test. Is there any trivial way to do it?

File: src/hypertrace/agent/init/__init__.py [hunk 11]
```diff
@@ -20,11 +20,10 @@
 
 class AgentInit:  # pylint: disable=R0902,R0903
     '''Initialize all the OTel components using configuration from AgentConfig'''
-    def __init__(self, agent: AgentConfig, init_console_only: bool = False):
+    def __init__(self, agentConfig: AgentConfig, init_console_only: bool = False):
```
Comment (jcchavezs) [inline src/hypertrace/agent/init/__init__.py:23]: Should this be following the naming conventions? Not sure if lint catched this.

File: agent-config.yaml (hunk 1 above)
Comment (nsahai) [reply agent-config.yaml:1]: @jcchavezs  we don't need this file and can be removed. I'll remove this file.

**Comment by jcchavezs** [comment]
@nsahai could you please rebase?

File: src/hypertrace/agent/config/agent-config.yaml (hunk 7 above)
Comment (rcbjBlueMars) [reply src/hypertrace/agent/config/agent-config.yaml:1]: This file sets unique configuration values that we are then checking in the unittest for validating that config file values are read correctly.

File: src/hypertrace/agent/config/__init__.py (hunk 6 above)
Comment (rcbjBlueMars) [reply src/hypertrace/agent/config/__init__.py:300]: Fixed.

File: src/hypertrace/agent/config/test_unittests.py (hunk 8 above)
Comment (rcbjBlueMars) [reply src/hypertrace/agent/config/test_unittests.py:1]: It was renamed to test_config.py.

File: src/hypertrace/agent/config/test_unittests.py (hunk 10 above)
Comment (rcbjBlueMars) [reply src/hypertrace/agent/config/test_unittests.py]: This is now done.

File: src/hypertrace/agent/config/test_unittests.py (hunk 9 above)
Comment (rcbjBlueMars) [reply src/hypertrace/agent/config/test_unittests.py]: This is fixed.

File: src/hypertrace/agent/init/__init__.py (hunk 11 above)
Comment (rcbjBlueMars) [reply src/hypertrace/agent/init/__init__.py]: An oversight. pylint seemed to miss it. It is now called agent_config.

File: src/hypertrace/agent/config/__init__.py (hunk 5 above)
Comment (rcbjBlueMars) [reply src/hypertrace/agent/config/__init__.py:106]: Fixed.

File: src/hypertrace/agent/__init__.py (hunk 4 above)
Comment (rcbjBlueMars) [reply src/hypertrace/agent/__init__.py:181]: Done.

File: src/hypertrace/agent/__init__.py (hunk 3 above)
Comment (rcbjBlueMars) [reply src/hypertrace/agent/__init__.py]: done.

**Comment by rcbjBlueMars** [comment]
All issues have been addressed.

File: src/hypertrace/agent/config/agent-config.yaml (hunk 7 above)
Comment (jcchavezs) [reply src/hypertrace/agent/config/agent-config.yaml:1]: Then I'd rather make it explicit in the name. We can address it in a next PR.
